### PR TITLE
Add additional scrape config params to PodMonitoring

### PR DESCRIPTION
**Pull request by fabxc**

### CODE REVIEW

**Comment by pintohutch** [comment]
Could you rebase this as well? I missed this was the child in a PR chain.

**Comment by fabxc** [comment]
Done

File: pkg/operator/operator.go [hunk 1]
```diff
@@ -74,8 +74,8 @@ const (
 
 	// The k8s Application, will be exposed as component name.
 	KubernetesAppName    = "app"
-	CollectorAppName     = "managed-prometheus-collector"
-	RuleEvaluatorAppName = "managed-prometheus-rule-evaluator"
+	CollectorAppName     = "managed_prometheus_collector"
```
Comment (pintohutch) [inline pkg/operator/operator.go:77]: I think your rebase may be stale here, on main this isn't changed: https://github.com/GoogleCloudPlatform/prometheus-engine/blob/main/pkg/operator/operator.go#L77

File: pkg/operator/apis/monitoring/v1alpha1/types.go [hunk 2]
```diff
@@ -562,17 +587,41 @@ func labelMappingRelabelConfigs(mappings []LabelMapping, prefix string) ([]*rela
...
+	TargetLabels TargetLabels `json:"targetLabels,omitempty"`
+	// Limits to apply at scrape time.
+	Limits *ScrapeLimits `json:"limits,omitempty"`
+}
+
+// ScrapeLimits limits applied to scraped targets.
+type ScrapeLimits struct {
+	// Maximum number of samples accepted within a single scrape.
+	Samples uint64 `json:"samples,omitempty"`
+	// Maximum number of labels accepted for a single sample.
```
Comment (pintohutch) [inline pkg/operator/apis/monitoring/v1alpha1/types.go:605]: Can we add a note about what happens if left unspecified?

File: pkg/operator/apis/monitoring/v1alpha1/types.go [hunk 3]
```diff
@@ -421,17 +423,40 @@ func (pm *PodMonitoring) endpontScrapeConfig(index int) (*promconfig.ScrapeConfi
...
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid proxy URL")
+	}
+	// Marshalling the config will redact the password, so we don't support those.
+	// It's not a good idea anyway and we will later support basic auth based on secrets to
+	// cover the general use case.
+	if _, ok := proxyURL.User.Password(); ok {
+		return nil, errors.New("passwords encoded in URLs are not supported")
+	}
+	// Initialize from default as encode/decide does not work correctly with the type definition.
```
Comment (pintohutch) [inline pkg/operator/apis/monitoring/v1alpha1/types.go:436]: typo: decide

File: pkg/operator/apis/monitoring/v1alpha1/types_test.go [hunk 4]
```diff
@@ -372,7 +405,12 @@ honor_timestamps: false
 scrape_interval: 10s
 scrape_timeout: 10s
 metrics_path: /metrics
-follow_redirects: false
+sample_limit: 1
+label_limit: 2
+label_name_length_limit: 3
+label_value_length_limit: 4
+proxy_url: ""
+follow_redirects: true
```
Comment (pintohutch) [inline pkg/operator/apis/monitoring/v1alpha1/types_test.go:413]: Where does `follow_redirects` get set?

**Review by pintohutch — APPROVED** [review]
Mostly nits and some questions. LGTM % fix stale rebase I commented on.

File: pkg/operator/apis/monitoring/v1alpha1/types_test.go (hunk 4 above)
Comment (fabxc) [reply pkg/operator/apis/monitoring/v1alpha1/types_test.go:413]: It's through the DefaultHTTPClientConfig.
That's one of the things that's broken and why we manually have to set the DefaultHTTPClientConfig. It doesn't have `omitempty` set on the upstream struct. Hence when encoding it explicitly sets `follow_redirects: false` and then no longer defaults correctly to `true` when decoding again.

File: pkg/operator/operator.go (hunk 1 above)
Comment (fabxc) [reply pkg/operator/operator.go:77]: Ah, I resolved the rebase incorrectly.

File: pkg/operator/operator.go [hunk 5]
```diff
@@ -75,7 +75,7 @@ const (
 	// The k8s Application, will be exposed as component name.
 	KubernetesAppName    = "app"
 	CollectorAppName     = "managed-prometheus-collector"
-	RuleEvaluatorAppName = "managed-prometheus-rule-evaluator"
+	RuleEvaluatorAppName = "managed-prometheus-rule_evaluator"
```
Comment (pintohutch) [inline pkg/operator/operator.go:78]: i think this one got past you :)

**Pull request merged**